R/check_exercise.R

```diff
@@ -0,0 +1,407 @@
+        exercise_eval <- !exercise_eval
+      }
+
+
+      # set the eval property as appropriate
+      options$eval <- exercise_eval
+      options$echo <- FALSE
+    }
+
+    # Add the options list to the chunk
+    usefull_chunks[[i]]$options <- options
+
+    # The exercise object will need a  label and the engine,
+    # we set them both here
+    usefull_chunks[[i]]$label <-  usefull_chunks[[i]]$params$label
+    usefull_chunks[[i]]$engine <-
+      usefull_chunks[[i]]$options$engine <- usefull_chunks[[i]]$params$engine %||% "r"
+
+  }
+
+  # Setting something to return that will be turned to FALSE
+  # if any check fail
+  all_passed <- TRUE
+
+  # Now that we have manipulated the chunks, we can build the
+  # exercise objects
+
+  for (chunk_ in chunks_){
+    # Restoring  knitr code stock
+    knitr::knit_code$restore()
+    # Grep all the related chunks (i.e setup, checker, etc)
+    all_related <- grep(chunk_, names(usefull_chunks), value = TRUE)
+
+    # If there is a setup chunk, we make sure it's the first of the list
+    if (any(grepl("setup", all_related))){
+      setup_ <- grepl("setup", all_related)
+      all_related <- c(
+        all_related[setup_],
+        all_related[!setup_]
+      )
+    }
+
+    # This function allow to grab a chunk based on
+    # its pattern
+    grab_chunk <- function(
+      pattern
+    ){
+      res <- usefull_chunks[
+        all_related[
+          grepl(
+            sprintf("%s-%s", chunk_, pattern),
+            all_related
+          )
+        ]
+      ]
+      if (length(res)){
+        return(res)
+      }
+      list(
+        list()
+      )
+    }
+
+    # Now we are building the exercise object, and it
+    # will be sent to evaluate_code()
+    #
+    # An exercise object needs the following elements:
+    # - exercise$label
+    # - exercise$code, which will be the solution code here
+    # - exercise$restore
+    # - exercise$timestamp
+    # - exercise$global_setup if any
+    # - exercise$setup if any
+    # - exercise$code_check if any
+    # - exercise$chunks that contains the setup chunks
+    # - exercise$check if any
+    # - exercise$engine
+    #
+
+    # We keep the  chunk we need: setup, and solution
+    chunks_needed <- list()
+
+    if (length(grab_chunk("setup")[[1]])){
+      chunks_needed[[1]] <- grab_chunk("setup")[[1]]
+    }
+    if (length(grab_chunk("solution")[[1]])){
+      chunks_needed[[
+        length(chunks_needed) + 1
+      ]] <- grab_chunk("solution")[[1]]
+    }
+    exercise_blt <- list(
+      # Getting the label
+      label = {
+        chunk_
+      },
+      code = {
+        paste0(grab_chunk("solution")[[1]]$code, "\n\n", collapse = "\n")
+      },
+      restore = {
+        FALSE
+      },
+      timestamp = {
+        as.numeric(Sys.time())
+      },
+      global_setup = {
+        paste0(setup_chunk$code, collapse = "\n")
+      },
+      setup = {
+        paste0(grab_chunk("setup")[[1]]$code, collapse = "\n")
+      },
```

Comment:
This is currently not used in `evaluate_exercise()`. Can set to NULL.

```suggestion
      setup = NULL,
```